confluent_kafka/schema_registry/protobuf.py

```diff
@@ -0,0 +1,465 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import io
+import sys
+import base64
+import struct
+from collections import deque
+
+from google.protobuf.message import DecodeError
+from google.protobuf.message_factory import MessageFactory
+
+from . import (_MAGIC_BYTE,
+               reference_subject_name_strategy,
+               topic_subject_name_strategy,)
+from .schema_registry_client import (Schema,
+                                     SchemaReference)
+from confluent_kafka.serialization import SerializationError
+
+# Converts an int to bytes (opposite of ord)
+# Python3.chr() -> Unicode
+# Python2.chr() -> str(alias for bytes)
+if sys.version > '3':
+    def _bytes(b):
+        """
+        Convert int to bytes
+
+        Args:
+            b (int): int to format as bytes.
+
+        """
+        return bytes((b,))
+else:
+    def _bytes(b):
+        """
+        Convert int to bytes
+
+        Args:
+            b (int): int to format as bytes.
+
+        """
+        return chr(b)
+
+
+class _ContextStringIO(io.BytesIO):
+    """
+    Wrapper to allow use of StringIO via 'with' constructs.
+
+    """
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+        return False
+
+
+def _create_msg_index(msg_desc):
+    """
+    Maps the location of msg_desc within a FileDescriptor.
+
+    Args:
+        msg_desc (MessageDescriptor): Message descriptor
+
+    Returns:
+        [int]: Message
```

Comment:
This is also a bit confusing, is this a pb or kafka Message? the module would do good here.

Reply:
it's also wrong it returns the message index which is neither a protobuf Message nor a Kafka Message. I'll prefix Message and add index